File: typescript-node/src/lookup.ts
import type { SkillPlus, QueryResult, ScanAck, ScanReport } from "skillplus";

/**
 * Steps 2 and 3 — get a report, scanning if there is not one yet.
 *
 * These are the two calls that make something happen; everything else in the SDK
 * reads what they produced. Choosing between them is a question about who is
 * waiting.
 */

/**
 * Step 2 — `query`: is there a report?
 *
 * The answer is binary. `status` is `"found"` (report attached) or
 * `"not_found"` (report is null). There is no third state to handle, and
 * failures arrive as thrown `SkillPlusError`s rather than as a status value —
 * so a `status` check can never accidentally swallow an outage.
 *
 * Three modes, in increasing order of how long they can take:
 */
export async function lookup(client: SkillPlus, repoUrl: string): Promise<QueryResult> {
  // 2a. Read only. Returns whatever exists right now and never triggers work.
  //     This is what the example uses, so that step 3 below is visibly the
  //     thing that starts a scan.
  const readOnly: QueryResult = await client.query({ repoUrl });

  // 2b. Read, and queue a scan if nothing exists. Still returns immediately:
  //     `not_found` with `scanning: true`, and the report is there next time.
  //
  //     This is the better default for a real detail page — one call instead of
  //     two, and the first visitor pays nothing. It is commented out here only
  //     because combining it with an explicit scan() below would queue the same
  //     scan twice and report a confusing "accepted, already in progress".
  //     Use ONE of the two, not both.
  //
  //   await client.query({ repoUrl, scanIfMissing: true });

  // 2c. Read, scanning if needed, and BLOCK until a completed report exists.
  //     `wait: true` implies scanIfMissing. This runs a full multi-agent audit,
  //     so it takes MINUTES — fine in CI or a batch job, never in a request
  //     handler.
  //
  //     The overload also narrows the type: with `wait: true` the result is
  //     `QueryFound`, so `.report` is non-null without a check.
  //
  //   const waited = await client.query({ repoUrl, wait: true });
  //   console.log(waited.report.verdict);  // no narrowing needed

  return readOnly;
}

/**
 * Step 3 — `scan`: ask for work to be done.
 *
 * Also two modes, and the return type differs between them.
 */
export async function startScan(client: SkillPlus, repoUrl: string): Promise<ScanAck> {
  // 3a. Fire-and-forget. Returns an acknowledgement, not a report.
  //
  //     `accepted: false` is NOT an error — it means the request was
  //     deduplicated, because a scan is already running or a fresh report
  //     already exists. Treating it as a failure and retrying is the usual bug
  //     here; the correct reading is "nothing more for you to do".
  const ack = await client.scan({
    repoUrl,

    // Scan a specific skill inside a multi-skill repository.
    // skillPath: "skills/my-skill",

    // Re-scan even when a fresh report exists. Without this, a repeat call is
    // deduplicated and returns accepted: false.
    // force: true,
  });

  return ack;
}

/**
 * 3b. Scan and wait — returns the ScanReport itself rather than an ack.
 *
 * Combined with `force: true`, the SDK does something worth knowing about: it
 * remembers the report that existed before, and will not accept that same old
 * report as the answer. Without that, a forced re-scan would return instantly
 * with the stale report and look like it had worked.
 *
 * Minutes, not seconds. Never call this from a request handler.
 */
export async function scanAndWait(client: SkillPlus, repoUrl: string): Promise<ScanReport> {
  return client.scan({
    repoUrl,
    wait: true,

    // How often to poll while waiting. Default 5_000.
    waitIntervalMs: 5_000,

    // Give up after this long and throw. Default 600_000 (10 minutes).
    waitTimeoutMs: 600_000,
  });
}

File: python/src/skillplus_example/lookup.py
"""Steps 2 and 3 — get a report, scanning if there is not one yet.

These are the two calls that make something happen; everything else in the SDK
reads what they produced. Choosing between them is a question about who is
waiting.
"""

from __future__ import annotations

from skillplus import QueryResult, ScanAck, ScanReport, SkillPlus


def lookup(client: SkillPlus, repo_url: str) -> QueryResult:
    """Step 2 — ``query``: is there a report?

    The answer is binary. ``status`` is ``"found"`` (``report`` attached) or
    ``"not_found"`` (``report`` is None). There is no third state, and failures
    arrive as a raised ``SkillPlusError`` rather than as a status value — so a
    status check can never accidentally swallow an outage.
    """
    # 2a. Read only. Returns whatever exists right now and never triggers work.
    #     This is what the example uses, so that step 3 is visibly the thing
    #     that starts a scan.
    result = client.query(repo_url)

    # 2b. Read, and queue a scan if nothing exists. Still returns immediately:
    #     not_found with scanning=True, and the report is there next time.
    #
    #     Better default for a real detail page — one call instead of two. It is
    #     commented out here only because combining it with an explicit scan()
    #     would queue the same scan twice. Use ONE of the two, not both.
    #
    #   client.query(repo_url, scan_if_missing=True)

    # 2c. Read, scanning if needed, and BLOCK until a completed report exists.
    #     wait=True implies scan_if_missing. This runs a full multi-agent audit,
    #     so it takes MINUTES — fine in a batch job, never in a request handler.
    #
    #   waited = client.query(repo_url, wait=True)
    #   print(waited.report.verdict)

    return result


def start_scan(client: SkillPlus, repo_url: str) -> ScanAck:
    """Step 3a — ``scan``, fire-and-forget. Returns an ack, not a report.

    ``accepted=False`` is NOT an error — it means the request was deduplicated,
    because a scan is already running or a fresh report already exists.
    Treating it as a failure and retrying is the usual bug here; the correct
    reading is "nothing more for you to do".
    """
    return client.scan(  # type: ignore[return-value]
        repo_url,
        # Scan one skill inside a multi-skill repository.
        # skill_path="skills/my-skill",
        #
        # Re-scan even when a fresh report exists. Without this a repeat call is
        # deduplicated and comes back accepted=False.
        # force=True,
    )


def scan_and_wait(client: SkillPlus, repo_url: str) -> ScanReport:
    """Step 3b — scan and wait. Returns the report itself rather than an ack.

    Combined with ``force=True`` the SDK does something worth knowing about: it
    remembers the report that existed before and refuses to accept that same old
    report as the answer. Without it, a forced re-scan would return instantly
    with the stale report and look like it had worked.

    Minutes, not seconds. Never call this from a request handler.
    """
    return client.scan(  # type: ignore[return-value]
        repo_url,
        wait=True,
        wait_interval=5.0,   # seconds between polls (default 5.0)
        wait_timeout=600.0,  # give up and raise after this long (default 600.0)
    )
